File: appAndroid/common-compose/src/main/java/com/gchristov/newsfeed/commoncompose/elements/AppError.kt
package com.gchristov.newsfeed.commoncompose.elements

import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Clear
import androidx.compose.material.icons.filled.Warning
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.gchristov.newsfeed.commoncompose.theme.Theme

data class BlockingError(
    val title: String,
    val subtitle: String,
    val retry: String
)

data class NonBlockingError(
    val title: String,
    val subtitle: String,
    val retry: String
)

fun Throwable.toUiBlockingError() = BlockingError(
    title = "Oups!",
    subtitle = "Something has gone wrong. Please try again.",
    retry = "Try again"
)

fun Throwable.toUiNonBlockingError() = NonBlockingError(
    title = "Oups",
    subtitle = "Something went wrong. Please try again.",
    retry = "Retry"
)

@Composable
fun AppBlockingError(
    blockingError: BlockingError,
    onRetry: () -> Unit
) {
    Box(
        modifier = Modifier.fillMaxSize(),
        contentAlignment = Alignment.Center
    ) {
        Column(
            modifier = Modifier.fillMaxWidth(0.5f),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(24.dp)
        ) {
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(4.dp)
            ) {
                AppText(
                    textAlign = TextAlign.Center,
                    text = blockingError.title,
                    style = Theme.typography.title,
                )
                AppText(
                    textAlign = TextAlign.Center,
                    text = blockingError.subtitle,
                    style = Theme.typography.subtitle,
                    color = Theme.contentColors.secondary
                )
            }
            AppButton(
                text = blockingError.retry,
                onClick = onRetry,
            )
        }
    }
}

@Composable
fun AppNonBlockingError(
    nonBlockingError: NonBlockingError,
    onRetry: () -> Unit,
    onDismiss: () -> Unit
) {
    Box(
        modifier = Modifier
            .fillMaxSize()
            .padding(8.dp),
        contentAlignment = Alignment.BottomCenter
    ) {
        AppSurface(elevation = 10.dp) {
            Row(modifier = Modifier.fillMaxWidth()) {
                AppIcon(
                    modifier = Modifier.padding(end = 16.dp, top = 4.dp),
                    imageVector = Icons.Filled.Warning,
                    tint = Theme.contentColors.warning
                )
                Column(
                    modifier = Modifier.weight(1f),
                    verticalArrangement = Arrangement.spacedBy(2.dp)
                ) {
                    AppText(
                        text = nonBlockingError.title,
                        style = Theme.typography.title,
                    )
                    AppText(
                        text = nonBlockingError.subtitle,
                        style = Theme.typography.subtitle,
                        color = Theme.contentColors.secondary
                    )
                    AppTextButton(
                        onClick = onRetry,
                        text = nonBlockingError.retry
                    )
                }
                AppIconButton(
                    modifier = Modifier.offset(x = 8.dp, y = (-8).dp),
                    icon = Icons.Filled.Clear,
                    onClick = onDismiss,
                    iconTint = Theme.contentColors.action
                )
            }
        }
    }
}

File: ios/common-swiftui/common-swiftui/elements/AppError.swift
import SwiftUI

public struct BlockingError {
    let title: String
    let subtitle: String
    let retry: String
    
    public init(
        title: String = "Oups!",
        subtitle: String = "Something has gone wrong. Please try again.",
        retry: String = "Try again"
    ) {
        self.title = title
        self.subtitle = subtitle
        self.retry = retry
    }
}

public struct NonBlockingError {
    let title: String
    let subtitle: String
    let retry: String
    
    public init(
        title: String = "Oups",
        subtitle: String = "Something went wrong. Please try again.",
        retry: String = "Retry"
    ) {
        self.title = title
        self.subtitle = subtitle
        self.retry = retry
    }
}

public struct AppBlockingErrorView: View {
    @EnvironmentObject var theme: Theme
    private let blockingError: BlockingError
    private let onRetry: () -> ()
    
    public init(
        blockingError: BlockingError,
        onRetry: @escaping () -> ()
    ) {
        self.blockingError = blockingError
        self.onRetry = onRetry
    }
    
    public var body: some View {
        GeometryReader { metrics in
            VStack {
                Spacer()
                HStack {
                    Spacer()
                    VStack(
                        alignment: .center,
                        spacing: 24
                    ) {
                        VStack(spacing: 4) {
                            AppText(
                                text: blockingError.title,
                                font: theme.typography.title
                            )
                            AppText(
                                text: blockingError.subtitle,
                                color: theme.contentColors.secondary,
                                font: theme.typography.subtitle,
                                textAlignment: .center
                            )
                        }
                        AppButton(
                            text: blockingError.retry,
                            onClick: onRetry
                        )
                    }.frame(maxWidth: metrics.size.width * 0.5)
                    Spacer()
                }
                Spacer()
            }
        }
    }
}

public struct AppNonBlockingErrorView: View {
    @EnvironmentObject var theme: Theme
    private let nonBlockingError: NonBlockingError
    private let onRetry: () -> ()
    private let onDismiss: () -> ()
    
    public init(
        nonBlockingError: NonBlockingError,
        onRetry: @escaping () -> (),
        onDismiss: @escaping () -> ()
    ) {
        self.nonBlockingError = nonBlockingError
        self.onRetry = onRetry
        self.onDismiss = onDismiss
    }
    
    public var body: some View {
        VStack {
            Spacer()
            AppSurface {
                HStack(alignment: .top) {
                    AppIcon(
                        imageName: "exclamationmark.triangle.fill",
                        tint: theme.contentColors.warning
                    )
                        .padding(.top, 4)
                        .padding(.trailing, 8)
                    VStack(
                        alignment: .leading,
                        spacing: 2
                    ) {
                        AppText(
                            text: nonBlockingError.title,
                            font: theme.typography.title
                        )
                        AppText(
                            text: nonBlockingError.subtitle,
                            color: theme.contentColors.secondary,
                            font: theme.typography.subtitle
                        )
                        AppTextButton(
                            text: nonBlockingError.retry,
                            onClick: onRetry
                        ).padding(.top, 8)
                    }
                    Spacer()
                    AppIconButton(
                        imageName: "xmark",
                        tint: theme.contentColors.action,
                        contentDescription: "Close"
                    ) {
                        onDismiss()
                    }.padding(.top, 4)
                }
            }
        }.padding(8)
    }
}

struct Error_Previews: PreviewProvider {
    static var previews: some View {
        //        BlockingErrorView(
        //            blockingError: BlockingError(),
        //            onRetry: {}
        //        )
        AppNonBlockingErrorView(
            nonBlockingError: NonBlockingError(),
            onRetry: {},
            onDismiss: {}
        )
    }
}
